### PR TITLE
Storage: Dashboard summary builder cleanup

### PR DESCRIPTION
**Pull request by ryantxu**
Extracted from #56348, and helps address some of the concerns with #56507

This updates the dashboard summary builder so it does not lookup datasources directly.  This will index exactly what is said in the docs, and a second step (and periodic process) can then resolve the links to real items.  eg:
image

This PR tries to cleanup some of the issues with `kind=panel`, by adding a new `kind=plugin` and setting that explicitly.  Still issues, but I think this is closer to expressing things better.

### CODE REVIEW

File: pkg/services/store/kind/dashboard/testdata/gdev-walk-graph-gradient-area-fills.json [hunk 1]
```diff
@@ -15,15 +15,17 @@
       "kind": "panel",
       "name": "Req/s",
       "URL": "/d/graph-gradient-area-fills.json/panel-tests-graph-gradient-area-fills?viewPanel=2",
+      "fields": {
+        "type": "graph"
+      },
       "references": [
         {
-          "kind": "ds",
-          "type": "default.type",
```
Comment (ryantxu) [inline pkg/services/store/kind/dashboard/testdata/gdev-walk-graph-gradient-area-fills.json:21]: @ArturWierzbicki -- NOTE the version used in bluge will still fill in the default values.

File: pkg/services/store/kind/geojson/summary.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/models"
+)
+
+func GetObjectKindInfo() models.ObjectKindInfo {
```
Comment (ryantxu) [inline pkg/services/store/kind/geojson/summary.go:12]: I could make a new PR for this... but seems ok here as we are still fleshing out the kind registry

File: pkg/models/object.go [hunk 3]
```diff
@@ -15,23 +15,36 @@ const (
...
+	StandardKindGeoJSON = "geojson"
+
 	// StandardKindQuery early development on panel query library
 	// the kind may need to change to better encapsulate { targets:[], transforms:[] }
 	StandardKindQuery = "query"
+
+	//----------------------------------------
+	// References are referenced from objects
+	//----------------------------------------
+
```
Comment (ArturWierzbicki) [inline pkg/models/object.go:38]: so iiuc, the `StandardReferenceX` is used for entities that:
- are referred to by entities stored in the object store
- are not stored in the object store themselves

I like this distinction - object store internal will have to treat these two types of references differently (e.g. graph resolution algorithm can't process the "external" refs at all).


I don't get the distinction between `StandardReferencePlugin` and `StandardReferenceType_RuntimeTransformer` - to me they look the same, and both should be on the same level, ie. either

```
- ds
- dashboard
- query
- png
- runtime
├── plugin
└── transformer
```

or..



```
- ds
- dashboard
- query
- png
- ext_plugin / runtime_plugin
- ext_transformer / runtime_transformer
```

Also - would we want to reuse those 'runtime' references for non-grafana entites? E.g. google cal, maps, sso team etc

File: pkg/services/store/kind/dashboard/testdata/gdev-walk-graph-shared-tooltips.json [hunk 4]
```diff
@@ -32,23 +34,27 @@
...
-          "type": "default.type",
-          "UID": "default.uid"
+          "kind": "ds"
         },
         {
-          "kind": "panel",
-          "type": "xychart"
+          "kind": "plugin",
+          "type": "panel",
+          "UID": "xychart"
```
Comment (ArturWierzbicki) [inline pkg/services/store/kind/dashboard/testdata/gdev-walk-graph-shared-tooltips.json:47]: ❤️

File: pkg/models/object.go (hunk 3 above)
Comment (ryantxu) [reply pkg/models/object.go:38]: I agree the taxonomy here is challenging -- but think this is progress in the right direction.

"runtime" is not great -- maybe "frontend"?  

plugin is big enough and has a clear 'type'

Flat seems attractive, but the list will get pretty long and has implied heiarchy anyway:

- panel-plugin
- datasource-plugin
- app-plugin
- influx-ds.  (datasource instance of type influx -- also requires the influx datasource plugin!)
- prometheus-ds
- ...



`StandardReferenceType_RuntimeTransformer` -- mainly a constant for transformer.  could/should move to a registry of the possible frontend types 🤷

Comment (ArturWierzbicki) [reply pkg/models/object.go:38]: I am not sure about `frontend` - I am pretty sure we could use this pattern for backend/non-grafana entities as well (like plugin?). Wdyt about `ExternalEntityReferenceX`, as in external to the object store/entity api? I think that's the most important distinction for us

**Comment by ryantxu** [comment]
@ArturWierzbicki -- I changed to `ExternalEntityReferenceX`

**Review by ArturWierzbicki — APPROVED** [review]

**Pull request merged**